### PR TITLE
[Product] Product variant generation refactoring

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Related tickets |  |
| License | MIT |

Based on #5068 and #5127

### CODE REVIEW

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        $smallBlueTShirt->hasOption($mediumSize)->willReturn(false);
+        $mediumBlueTShirt->hasOption($blueColor)->willReturn(true);
+        $mediumBlueTShirt->hasOption($mediumSize)->willReturn(true);
+
+        $checkMediumBlueTShirt->getOptions()->willReturn([$blueColor, $mediumSize]);
+
+        $this->checkParity($checkMediumBlueTShirt, $tShirt)->shouldReturn(true);
+    }
+
+    function it_returns_null_if_variable_does_not_contain_variable_with_same_options_set(
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: `returns_null`?

Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: does_not_contain_varia**nt**

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        OptionInterface $size
+    ) {
+        $tShirt->getVariants()->willReturn([$checkMediumBlueTShirt]);
+        $tShirt->getOptions()->willReturn([$color, $size]);
+
+        $checkMediumBlueTShirt->getOptions()->willReturn([$blueColor, $mediumSize]);
+
+        $this->checkParity($checkMediumBlueTShirt, $tShirt)->shouldReturn(false);
+    }
+}
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: Missing spec for assertion.

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        OptionValueInterface $mediumSize,
+        OptionInterface $color,
+        OptionInterface $size
+    ) {
+        $tShirt->getVariants()->willReturn([$checkMediumBlueTShirt]);
+        $tShirt->getOptions()->willReturn([$color, $size]);
+
+        $checkMediumBlueTShirt->getOptions()->willReturn([$blueColor, $mediumSize]);
+
+        $this->checkParity($checkMediumBlueTShirt, $tShirt)->shouldReturn(false);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: When the variant is the same as already existing one on a product shouldn't we return `true`?

File: src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php [hunk 4]
```diff
@@ -84,14 +93,62 @@ function it_generates_variants_for_every_value_of_an_objects_single_option(
...
+            ['black1', 'white2', 'red3'],
+        ]);
+
+        $productVariable->getMasterVariant()->willReturn($masterVariant);
+        $variantsParityChecker->checkParity($permutationVariant, $productVariable)->willReturn(true);
+
+        $variantFactory->createNew()->willReturn($permutationVariant);
+        $permutationVariant->setObject($productVariable)->shouldBeCalled();
+        $permutationVariant->setDefaults($masterVariant)->shouldBeCalled();
+        $permutationVariant->addOption(Argument::type('Sylius\Component\Variation\Model\OptionValue'))->shouldBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php]: `::class`

File: src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php [hunk 5]
```diff
@@ -84,14 +93,62 @@ function it_generates_variants_for_every_value_of_an_objects_single_option(
...
+        VariantInterface $permutationVariant,
+        VariantsParityCheckerInterface $variantsParityChecker
+    ) {
+        $productVariable->hasOptions()->willReturn(true);
+
+        $productVariable->getOptions()->willReturn([$colorOption]);
+
+        $colorOption->getValues()->willReturn([$blackColor, $whiteColor, $redColor]);
+
+        // Stubbing `OptionValue` instead of `OptionValueInterface` in order to stub `getId` method.
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php]: `OptionValueInterface` extends `ResourceInterface` which has `getId` method defined. You should use interface.

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php (hunk 3 above)
Comment (lchrusciel) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: For validator logic we should skip an object if it is placed in the same place in memory. But in every other use case, you are right, that we should return true. But how to name this two slightly different methods?

Comment (pamil) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: `TShirt` / `tShirt` -> `Tshirt` / `tshirt` ?

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        $tShirt->getOptions()->willReturn([$color, $size]);
+
+        $checkMediumBlueTShirt->getOptions()->willReturn([$blueColor]);
+
+        $this
+            ->shouldThrow(
+                new \InvalidArgumentException(
+                    'Number of set option values should be equal to number of available options.'
+                )
+            )->during('checkParity', [$checkMediumBlueTShirt, $tShirt]);
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: Move semicolon to next line.

File: src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php [hunk 7]
```diff
@@ -74,24 +82,72 @@ function it_generates_variants_for_every_value_of_an_objects_single_option(
...
+
+        $blackColor->getId()->willReturn('black1');
+        $whiteColor->getId()->willReturn('white2');
+        $redColor->getId()->willReturn('red3');
+
+        $setBuilder->build([
+            ['black1', 'white2', 'red3'],
+        ])->willReturn([
+            ['black1', 'white2', 'red3'],
+        ]);
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php]: Won't it fit in one line?

File: src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php [hunk 8]
```diff
@@ -74,24 +82,72 @@ function it_generates_variants_for_every_value_of_an_objects_single_option(
...
-        VariantInterface $masterVariant, VariantInterface $permutationVariant
+        OptionInterface $colorOption,
+        OptionInterface $sizeOption,
+        OptionValueInterface $blackColor,
+        OptionValueInterface $whiteColor,
+        OptionValueInterface $redColor,
+        OptionValueInterface $smallSize,
+        OptionValueInterface $mediumSize,
+        OptionValueInterface $largeSize,
+        VariantInterface $masterVariant, VariantInterface $permutationVariant,
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/spec/Generator/VariantGeneratorSpec.php]: ``VariantInterface $permutationVariant`` should be placed in next line.

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    function it_matches_if_given_variable_contains_similar_variant(
+        VariableInterface $tShirt,
+        VariantInterface $checkMediumBlueTShirt,
+        VariantInterface $smallBlueTShirt,
+        VariantInterface $mediumBlueTShirt,
+        OptionValueInterface $blueColor,
+        OptionValueInterface $mediumSize,
+        OptionInterface $color,
+        OptionInterface $size
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: How about putting variables in alphabetic order?

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        $tShirt->getVariants()->willReturn([$checkMediumBlueTShirt]);
+        $tShirt->getOptions()->willReturn([$color, $size]);
+
+        $checkMediumBlueTShirt->getOptions()->willReturn([$blueColor, $mediumSize]);
+
+        $this->checkParity($checkMediumBlueTShirt, $tShirt)->shouldReturn(false);
+    }
+
+
+    function it_throws_an_exception_if_number_of_possible_options_on_variable_and_number_of_options_values_on_variant_are(
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: ``.._are`` - are what? 🐫 😄

File: src/Sylius/Bundle/VariationBundle/spec/Validator/VariantCombinationValidatorSpec.php [hunk 11]
```diff
@@ -37,20 +39,28 @@ function it_is_a_constraint_validator()
         $this->shouldImplement(ConstraintValidator::class);
     }
 
-    function it_should_not_add_violation_if_variant_is_master(VariantInterface $variant)
-    {
+    function it_should_not_add_violation_if_variant_is_master(
+        $context,
```
Comment (Zales0123) [inline src/Sylius/Bundle/VariationBundle/spec/Validator/VariantCombinationValidatorSpec.php]: Have you thought about typehinting all parameters?

File: src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php (hunk 3 above)
Comment (lchrusciel) [inline src/Sylius/Component/Variation/spec/Checker/VariantsParityCheckerSpec.php]: TShirt seems to be more appropriate, as there are 2 words(letters?) at the end.

**Comment by lchrusciel** [comment]
Closing as it will be a part of bringing back product variant generation

**Pull request closed without merging**